lib/charms/prometheus_k8s/v0/prometheus_remote_write.py

```diff
@@ -0,0 +1,124 @@
+"""
+
+from typing import List, Optional, Union
+
+from ops.charm import CharmBase
+from ops.framework import Object
+from ops.model import BlockedStatus
+
+
+class PrometheusRemoteWriteConsumer(Object):
+    """A prometheus remote write consumer."""
+
+    def __init__(self, charm: CharmBase, relation_name: str):
+        """A prometheus remote write consumer.
+
+        Args:
+            charm: The charm object that instantiated this class.
+            relation_name: The relation name as defined in metadata.yaml.
+        """
+        super().__init__(charm, relation_name)
+        self._relation_name = relation_name
+
+    @property
+    def endpoints(self) -> List[str]:
+        """A list of remote write endpoints.
+
+        Returns:
+            A list of remote write endpoints.
+        """
+        endpoints = []
+        for relation in self.model.relations[self._relation_name]:
+            for unit in relation.units:
+                # If external-address is provided use that, else use ingress-address
+                if (address := relation.data[unit].get("address")) is None:
+                    continue
+                if (port := relation.data[unit].get("port")) is None:
+                    continue
+                endpoints.append(f"http://{address}:{port}/api/v1/write")
+        return endpoints
+
+    @property
+    def configs(self) -> list:
+        """A config object ready to be dropped in to a prometheus config file.
+
+        Returns:
+            A list of remote_write configs.
+        """
+        return [{"url": endpoint} for endpoint in self.endpoints]
+
+
+class PrometheusRemoteWriteProvider(Object):
+    """A prometheus remote write provider."""
+
+    def __init__(self, charm: CharmBase, relation_name: str):
+        """A prometheus remote write provider.
+
+        Args:
+            charm: The charm object that instantiated this class.
+            relation_name: The relation name as defined in metadata.yaml.
+        """
+        super().__init__(charm, relation_name)
+        self._charm = charm
+        self._relation_name = relation_name
+
+    def set_endpoint(
+        self,
+        port: Union[str, int],
+        address: Optional[str] = None,
+        ingress_relation: Optional[str] = None,
+        ingress_address: Optional[str] = None,
+    ) -> None:
+        """Set the address and port on which you will serve prometheus remote write.
+
+        If address is provided, all automatic logic will be skipped and the address provided will
+        be used. Otherwise the method will attempt to use the ingress information if it is
+        provided. Finally it will fall back to a predictable hostname if neither of the other two
+        methods worked.
+
+        Args:
+            port: The port number
+            address: The address of the remote write server
+            ingress_relation: The name of the ingress relation to use
+            ingress_address: The ingress address
```

Comment:
Why does `ingress_address` need to be a method argument when `ingress_relation` is already an optional one. For example if `ìngress_relation` is provided can it not be used to determine the `ingress_address` ?

Reply:
The relation can't give you the address. For example, in Prometheus, how do I get `_external_hostname` from the relation object?

Reply:
OK. Again the discussion here may be related to discussions above regarding library objects responding to events and setting endpoint transparently without the charm needing to invoke a method to set endpoints. If we adopt this approach then the arguments to this method may change. Something like "external hostname" can certainly be a constructor argument to this library object. There may be other options here for you to explore.